### ISSUE TITLE
Move actions before interaction profiles

### ISSUE THREAD

**Issue by rbessems** [KhronosGroup/OpenXR-Tutorials#32]
Move actions before interaction profiles. From a game developer perspective actions are the hero, understanding actions is the abstraction your game uses without caring what's behind it, thats the important bit.

Define good actions, build your game based on those actions.

Than interaction profiles are just a means to map those actions to controllers you're using.

Interaction profiles are just how you customize exposing those actions on specific controllers
getting your actions (and sub-action paths) and action sets  right is critical

**Comment by BastiaanOlij** [comment]
This for me is the main thing, actions and action sets are the heroes of the interaction system, the rest plays second fiddle. If we can make it clear what actions are and what a good setup is, and why and how you should use action sets and then go and show how you implement them, that will be important.

The biggest issue I see here is that the tutorial application likely doesn't need a complex action setup. It probably only needs one action set, so it'll be a challenge to get the point across.

Interaction profiles are just a means by which to map actions to actual inputs.

**Comment by rvkennedy** [comment]
I actually already did this in my first version of Ch4. I changed it back because the order in the [RFP](https://members.khronos.org/document/dl/29789) is: Spaces, then Interaction System, then Actions, then checking if a controller is connected.

So I wanted to discuss it before rearranging. @rbessems should we discuss this with the group or are you happy for me to make the change?

@BastiaanOlij what I think we can do is create an ActionSet with a cross-section of the action types, and have some visual response to each input. At the moment we read a bool, a float and a pose. But the float is from a boolean input because there are no float actions in the Khr Simple Controller Profile.

I must confess one are of the spec that I don't understand is that we have to specify the profile path always, so that multiple profiles can have the same user/component path, but there is no generic way to specify  the latter without choosing a specific profile.

**Comment by rbessems** [comment]
@rvkennedy absolutely go ahead and change the order, the RFP chapters are more guidance than gospel, rearrange what is needed to be a good tutorial.

**Comment by BastiaanOlij** [comment]
@rvkennedy yes, the simple controller is TOO simple. That is a discussion for another day :) 

Think of it this way. First you define all your actions and action sets, that is game facing, your game only needs to know about the actions and how to read them and thus not care what hardware is behind it. So a good design of an action map that serves your game well, is the most important thing for the game developer to do.

An interaction profile then maps 1 type of controller to those actions. So you're not linking `*/input/trackpad` to an action. You are linking that path for a specific controller. And yes that seems arduous because some paths exist on all controllers, but there are also many paths that only exist on 1 controller, or where you want to map something different for a given controller.

The best examples here are from the "old" days in VR when we just had the Vive, Rift and WMR headsets. The Vive only had a trackpad, the Rift only had a joystick, WMR had both.
Games developed on Vive favoured the trackpad, and thus on WMR the joystick is unused
Games developed on Quest favoured the joystick, and thus on WMR the trackpad is unused

Now here is the crux. The way OpenXR was designed is that you only have to specify one interaction profile, the one you as a developer use. You can specify multiple, but the assumption is that you only provide interaction profiles for hardware you've actually tested with.

So lets for argument say you had an HTC Vive and develop a game with that headset, you only specify the interaction profile for the HTC Vive, you don't go play a guessing game for how you'd map all the other controllers.

According to the OpenXR spec, the XR runtime should first check if an interaction profile exists for the hardware used, and then apply those bindings as provided, else it should find an interaction profile for the next best thing and do a mapping on those bindings, every runtime should support the simple controller as the ultimate fallback for this.

The problem here is that many XR runtimes ignore(d) this so the experience to the end user was less then stellar if bindings for their hardware aren't available. Some did a stellar job, some a dismal job.
In todays world due to the popularity of the Quest, most runtimes will fall back on mappings provided for the quest and do a good job remapping those.

Now I'm writing this all for context, obviously in a tutorial we want to assume runtimes play by the rules and we want to stress that the intention is to provide interaction profiles with bindings to your actions for the hardware you work on only.

**Comment by BastiaanOlij** [comment]
So as an example, let's say we define our action sets as follows:

```
- Action set "Menu"
  - Aim (Pose)
  - Select (Bool)
  - Cancel (Bool)

- Action set "Game"
  - Aim (Pose)
  - Grab (Bool)
  - Trigger (Bool)
  - Open Menu (Bool)
```

I'm missing alot of things here but just to keep it simple for this explination/suggestion. Also all actions here can be bound to either hand, as we'll see later on Aim, Select, Grab and Trigger will always be mapped to both hands, Cancel and Open Menu will be mapped to a specific input on one hand. You do this by querying subpaths, so when an action is bound to both hands, you can query that action for each hand individually.

So the action set "Menu" is active when we open our pause menu, we will position laser pointers based on each hands pose action, and with that we can aim at menu options. When the Select action is triggered, it will activate whatever button that hands laser pointer is pointing at. I usually do something funky like only show the laser pointer on which Select was last pressed, so people can easily switch between left or right handed interactions without needing to go into rebinding menus (that don't exist on any platform) or switch between left handed and right handed action sets, easy peasy.

If the cancel is triggered we simply close the menu.

The action set "Game" is active when the menu is closed. Again the aim poses are used to position analogues for hands, the user can pick up items by holding a hand close to that item and performing the grab action. Again this works for both hands, if the grab action is triggered on the left hand, the left hand picks up whatever item is closest to it, unless nothing is close to it, that nothing happens.
When the Grab action turns back to false (button is released), whatever the player is holding in that hand is dropped.

When the trigger action is triggered, than whatever item is held by the player in the related hand, is activated. If its a gun, it shoots, if its something else, it does whatever is appropriate for that item.

When the Open Menu action is triggered, we pauze our game and open the menu.

And voila, without ever talking about what hardware the user is using, we've defined the game loop for our user and coded our entire game.

So lets add our interaction profiles. Lets assume we only have access to an HTC vive, so we create an interaction profile for that first:

```
- /interaction_profiles/htc/vive_controller
 - Menu/Aim <= /user/hand/left/input/aim/pose, /user/hand/right/input/aim/pose
 - Menu/Select <= /user/hand/left/input/trigger/click, /user/hand/right/input/trigger/click
 - Menu/Cancel <= /user/hand/right/input/menu/click
 - Game/Aim <= /user/hand/left/input/aim/pose, /user/hand/right/input/aim/pose
 - Game/Grab <= /user/hand/left/input/squeeze/click, /user/hand/right/input/squeeze/click
 - Game/Trigger <= /user/hand/left/input/trigger/click, /user/hand/right/input/trigger/click
 - Menu/Open Menu <= /user/hand/right/input/menu/click
```

So obviously at this point we'd want to also add a fallback to our simple controller... Well even though our action map has very basic functions, the simple controller is already too simple, so thats a bad idea. I guess our simple game is already too complex, and we haven't even added locomotion... sigh.
Mind you, beat sabers action map would suffice, its basically Aim + Select, no more, just those two actions.
That might be an argument for creating such a game as our tutorial game, I dunne know.

We'd need to talk about that in a wider group I suppose.

However as I said most runtimes have recognised this for ages and most now use the Touch profile as a fallback, so we could suggest that this profile should be added. But it should be underlined that it is not the intention for a game to deliver profiles for all known interaction profiles.

Be that as it may, our Touch profile would look something like:

```
- /interaction_profiles/oculus/touch_controller
 - Menu/Aim <= /user/hand/left/input/aim/pose, /user/hand/right/input/aim/pose
 - Menu/Select <= /user/hand/left/input/trigger/value, /user/hand/right/input/trigger/value
 - Menu/Cancel <= /user/hand/left/input/menu/click
 - Game/Aim <= /user/hand/left/input/aim/pose, /user/hand/right/input/aim/pose
 - Game/Grab <= /user/hand/left/input/squeeze/value, /user/hand/right/input/squeeze/value
 - Game/Trigger <= /user/hand/left/input/trigger/value, /user/hand/right/input/trigger/value
 - Menu/Open Menu <= /user/hand/left/input/menu/click
```
Note that for touch both squeeze and trigger don't support click, but both support value (float) and both will be converted to bool just fine.
Also note that our menu button is only available on the left controller. A better suggestion may be using the A and B buttons.

Hope that helps.

**Comment by rbessems** [comment]
done
